dask_cuda/explicit_comms/dataframe/shuffle.py

```diff
@@ -5,201 +5,239 @@
+    partition ID maps to exactly one dataframe.
 
     Parameters
     ----------
-    in_parts: list of dict of dataframes
-        List of dataframe groups that need to be shuffled.
-    rank_to_out_part_ids: dict
-        dict that for each worker rank specifices a list of partition IDs that
-        worker should return. If the worker shouldn't return any partitions,
-        it is excluded from the dict.
+    df_meta: DataFrame
+        An empty dataframe matching the expected output
+    dfs: dict of dataframes
+        The dataframes to spilt given as a map of stage keys to dataframes
+    column_names: list of strings
+        List of column names on which we want to split.
+    npartitions: int or None
+        The desired number of output partitions.
     ignore_index: bool
-        Ignore index during shuffle.  If ``True``, performance may improve,
+        Ignore index during shuffle.  If True, performance may improve,
         but index values will not be preserved.
-    concat_dfs_of_same_output_partition: bool
-        Concatenate all dataframes of the same output partition.
+    proxify: callable
+        Function to proxify object.
 
     Returns
     -------
-    rank_to_out_parts_list: dict of list of list of DataFrames
-        Dict that maps each worker rank to its output partitions.
+    dict of DataFrames
+        Mapping from partition ID to dataframe.
     """
 
-    out_part_id_to_dataframes = defaultdict(list)  # part_id -> list of dataframes
-    for bins in in_parts:
-        for k, v in bins.items():
-            out_part_id_to_dataframes[k].append(v)
-        del bins
-
-    # Create mapping: rank -> list of [list of dataframes]
-    rank_to_out_parts_list: Dict[int, List[List[DataFrame]]] = {}
-    for rank, part_ids in rank_to_out_part_ids.items():
-        rank_to_out_parts_list[rank] = [out_part_id_to_dataframes[i] for i in part_ids]
-    del out_part_id_to_dataframes
-
-    # Concatenate all dataframes of the same output partition.
-    if concat_dfs_of_same_output_partition:
-        for rank in rank_to_out_part_ids.keys():
-            for i in range(len(rank_to_out_parts_list[rank])):
-                if len(rank_to_out_parts_list[rank][i]) > 1:
-                    rank_to_out_parts_list[rank][i] = [
-                        concat(
-                            rank_to_out_parts_list[rank][i], ignore_index=ignore_index
-                        )
-                    ]
-    return rank_to_out_parts_list
-
-
-async def local_shuffle(
+    # Grouping each input dataframe, one part for each partition ID.
+    dfs_grouped: List[Dict[int, DataFrame]] = []
+    while dfs:
+        dfs_grouped.append(
+            proxify(
+                single_shuffle_group(
+                    # pop dataframe in any order, to free staged memory ASAP
+                    dfs.popitem()[1],
+                    column_names,
+                    npartitions,
+                    ignore_index,
+                )
```

Comment:
So this does a round of communication (one for each input partition)?

Reply:
No communication here, this is creating the local shuffle groups